vignettes/simplify_function.Rmd

```diff
@@ -1,28 +1,121 @@
+
+You can also write your own custom simplify function to clean up the prediction results. For example, below we define a custom function that obtains the probabilities from the list of prediction results with keys being the prediction keys pre-defined for this canned estimator. In this case, `linear_classifier` has logits, probabilities, prediction classes, logistics, etc. pre-defined. We can see a full list of available predictions keys by calling `prediction_keys()`. 
+
+```{r}
+simplify_fn <- function(predictions) {
+  lapply(predictions, function(x) x$probabilities)
+}
+
+model %>% predict(mtcars_input_fn(mtcars[1:2, ]), simplify = simplify_fn)
+```
+Note again that we only make predictions for the first two rows of the data since it's too verbose to fit in this page.
+
+## Simplify Evaluation Results
+
+Similarly, the evaluation results can be simplified as well. Here's how the raw evaluation results look like without simplification by providing `simplify = FALSE`. 
+
+```{r}
+model %>% evaluate(mtcars_input_fn(mtcars[1, ]), simplify = FALSE)
 ```
+It contains a list of evaluation results for the metrics pre-defined for this particular canned estimator, such as accuracy, average loss, AUC, etc. The evaluation result is pretty verbose without simplification. 
 
-## Including Plots
+### Default Simplifier
 
-You can also embed plots, for example:
+Now let's take a look at the simplified results by providing `simplify = TRUE`:
 
-```{r pressure, echo=FALSE}
-plot(pressure)
+```{r}
+model %>% evaluate(mtcars_input_fn(test_data), simplify = TRUE)
 ```
+The evaluation results now look compact.  
+
+### Custom Simplifier for Evaluation Result
+
+We can also apply a custom simplifier on the evaluation result like the following to obtain the AUC score:
 
-Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
+```{r}
+simplify_fn <- function(results) {
+  results$auc 
+}
+
+model %>% evaluate(mtcars_input_fn(test_data), simplify = simplify_fn)
+```
+
+
+## Writing Simplify Function for Custom Estimators
+
+So far we've only illustrated how to simplify the results for canned estimator, in particular, `linear_classifier`. Custom estimators do not simplify results by default unless you provide a custom simplify function. The way to write them is almost identical to canned estimators. The predictions and evaluations results are highly dependent on what's defined in the `model_fn` so you need to pay attention to them when writing the simplify function. 
+
+For example, below is a snippet in the `model_fn` that defines the behavior for prediction mode (note the `mode == mode_keys()$PREDICT`). `predictions` is being returned during this phase and is a list of keyed items for raw prediction classes and probabilities.
+
+
+```{r eval=FALSE}
+predictions <- list(
+    class = tf$argmax(logits, 1L),
+    prob = tf$nn$softmax(logits))
+if (mode == mode_keys()$PREDICT) {
```

Comment:
I think we should be using straight character literals here rather than the enum style.

Reply:
How about using `mode_keys()$PREDICT` for development but using "infer" - (straight character literal) - in tutorials? This way it's easier for users and easier for maintenance.